### PR TITLE
Tabular: Added ImagePredictorModel

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added ImagePredictorModel

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1041-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/3/index.html

File: core/src/autogluon/core/features/types.py
Comment (jwmueller) [inline core/src/autogluon/core/features/types.py:35]: Might _path be more appropriate suffix than _url?

**Comment by jwmueller** [comment]
Can FeatureMetadata docs include somewhere: 

For descriptions of each special feature-type, see:  `autogluon.core.features.types`

File: core/src/autogluon/core/features/types.py
Comment (jwmueller) [reply core/src/autogluon/core/features/types.py:35]: `image_url` is missing from list of possible special types in FeatureMetadata documentation

**Comment by jwmueller** [comment]
Unrelated to this PR but can FeatureMetadata docs include a basic usage example specifying metadata and then calling `fit()`?  Without this, the role of the class may be unclear to user.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py
Comment (jwmueller) [inline features/src/autogluon/features/generators/auto_ml_pipeline.py:45]: Is this argument actually needed? why would a user specify `image_url` in metadata but not want to leverage the images?

File: core/src/autogluon/core/features/types.py
Comment (jwmueller) [reply core/src/autogluon/core/features/types.py:35]: Niche use-case but may want to clarify:  how user can leverage both image and its file-path both as predictive features

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: to standardize, should it be: `ImageNeuralNetwork` like for Text?

File: tabular/tests/unittests/models/test_image_predictor.py
Comment (jwmueller) [inline tabular/tests/unittests/models/test_image_predictor.py:14]: maybe also test setting `predictor=None` and reloading it before inference

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Does this post-processing really need to be so complex?

@zhreshold  Should the issue be fixed in autogluon.vision/gluonCV first before merging this PR?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:119]: should we raise clear error message if regression?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:28]: why None instead of boolean?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: what happens with HPO for this model?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:78]: can doc be clearer about what X/X_val must look like?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:143]: Should this error be moved earlier as well for `fit()`?

**Review by jwmueller — CHANGES_REQUESTED** [review]
Looks pretty good, mostly have minor suggestions. Major issue is the post-processing code needed at the end of `predict_proba()`

File: tabular/tests/unittests/models/test_image_predictor.py
Comment (jwmueller) [inline tabular/tests/unittests/models/test_image_predictor.py:1]: Can we have a similar script in `examples/`, so it can be shared with folks that want to test out this functionality?

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Ideally, gluoncv should just return the normal pred_proba format. I don't know why it is in the format it is, but it takes a **_lot_** of work to convert it to the standard format.

Comment (jwmueller) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Right I guess @zhreshold can tell us what's best for now regarding this issue.

Comment (zhreshold) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: See #1044, if the post-processing isn't a blocker then you could remove these once #1044 get merged.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (sxjscience) [inline tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:124]: We may need to call the following to avoid the case that some rare classes only appear in the validation set:

```python
if X_val is None:
    classes = list(y.unique())
else:
    classes = list(pd.concat([y, y_val]).unique())
```

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: I think rather Text should be TextPredictor. This makes it clear that the model is coming from the other predictors within AutoGluon, rather than something custom and specific to Tabular. In future, I'll also add `TabularPredictorModel` as a model type.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py
Comment (Innixma) [reply features/src/autogluon/features/generators/auto_ml_pipeline.py:45]: possibly for debugging, I agree it isn't really necessary, but doesn't hurt.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:162]: Will wait for #1044 to merge before addressing in this PR

File: tabular/tests/unittests/models/test_image_predictor.py
Comment (Innixma) [reply tabular/tests/unittests/models/test_image_predictor.py:14]: That is tested elsewhere, since we know predictor doesn't persist models in memory be default.

File: core/src/autogluon/core/features/types.py
Comment (Innixma) [reply core/src/autogluon/core/features/types.py:35]: renamed to `image_path`. Regarding using path as both text and image feature at the same time: Let's leave that for later, seems really niche.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:134]: added

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: renamed to `image_predictor`, this encompasses the entire predictor so it shouldn't be specific.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: removed docs, this was copy/pasted from Tabular's text predictor model.

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Good point, added

File: core/src/autogluon/core/features/types.py
Comment (Innixma) [reply core/src/autogluon/core/features/types.py]: added `image_path` to FeatureMetadata docs

**Comment by Innixma** [comment]
> Can FeatureMetadata docs include somewhere:
> 
> For descriptions of each special feature-type, see: `autogluon.core.features.types`

Added

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Added extended description

**Comment by szha** [comment]
Job PR-1041-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/4/index.html

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: Currently image model would fail unless a search space was provided by user (same as TextNN currently)

File: tabular/tests/unittests/models/test_image_predictor.py
Comment (Innixma) [reply tabular/tests/unittests/models/test_image_predictor.py:1]: Would prefer to get a golden dataset example for multimodality, we currently don't have one besides petfinder and petfinder is somewhat large for an example.

**Comment by szha** [comment]
Job PR-1041-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/5/index.html

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py
Comment (jwmueller) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: @sxjscience Can we standardize this in the future?
I also think `TextPredictor` might be more clear name than `TextNeuralNetwork`.

Comment (sxjscience) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:86]: Sounds good to me.

**Comment by szha** [comment]
Job PR-1041-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/6/index.html

File: tabular/src/autogluon/tabular/trainer/model_presets/presets.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/trainer/model_presets/presets.py:84]: Updated to TextPredictor

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Code block is now removed, no longer necessary after #1044

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py:1]: Added default HPO search space, but we probably want to have a better search space in future

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Removed this, it was copy-paste from TextPredictorModel

File: tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/image_prediction/image_predictor.py]: Thanks, added `len(pickle.dumps(self.model._classifier, pickle.HIGHEST_PROTOCOL))`

**Comment by szha** [comment]
Job PR-1041-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/9/index.html

**Comment by Innixma** [comment]
@zhreshold @sxjscience @jwmueller Addressed comments, ready for 2nd round of review.

File: features/src/autogluon/features/generators/auto_ml_pipeline.py
Comment (jwmueller) [inline features/src/autogluon/features/generators/auto_ml_pipeline.py:47]: "url path" -> "string path"

File: core/src/autogluon/core/features/feature_metadata.py
Comment (jwmueller) [inline core/src/autogluon/core/features/feature_metadata.py:181]: might be helpful if docstring provides a basic example use-case.

**Review by jwmueller — APPROVED** [review]
looks good, just had minor remaining comments

**Review by sxjscience — APPROVED** [review]
I’ve tried and it works well.

File: core/src/autogluon/core/features/feature_metadata.py
Comment (Innixma) [reply core/src/autogluon/core/features/feature_metadata.py]: added

File: features/src/autogluon/features/generators/auto_ml_pipeline.py
Comment (Innixma) [reply features/src/autogluon/features/generators/auto_ml_pipeline.py]: updated

**Comment by szha** [comment]
Job PR-1041-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1041/10/index.html

**Pull request merged**
